[macOS] Implement scheduling logic for default browser and add to dock prompts

alessandroboron opened the pull request:
Task/Issue URL: https://app.asana.com/1/137249556945/project/72649045549333/task/1210257532277816
Tech Design URL: https://app.asana.com/1/137249556945/project/481882893211075/task/1210257532277827
CC:

### Description
This PR implements the scheduling logic for presenting the 'Set as Default’ (SAD) browser and Add to dDock (ATT = Add to Taskbar) prompts.
Scheduling rules are defined in this [task](https://app.asana.com/1/137249556945/project/1206329551987282/task/1210226206083041?focus=true) 

### Testing Steps

**Prerequisites:**
- Run ./macOS/clean-app.sh debug|review (depending if testing a debug or ad-hoc build)
- Return `true` in `isDefaultBrowserAndDockPromptFeatureEnabled:Bool` defined in `DefaultBrowserAndDockPromptFeatureFlagger.swift`
- Set default values for prompt timing to low value (e.g. 3) in `DefaultBrowserAndDockPromptFeatureSettings defaultValue: Int` defined in `DefaultBrowserAndDockPromptFeatureFlagger.swift`

**Scenario 1 - Show Prompts Depending on Time Schedule**
1. Install the app, ensuring that the prerequisites are followed.
2. Open System Settings and click ‘Date & Time’ under ‘General'. 
3. Disable ‘Set time and date automatically’ to be able to override the current time.
4. Set a date in the future for which the SAD/ATT popover should not show.
5. Open a new window. 
6. Ensure that no SAD/ATT popover is prompted.
7. Set a date in the future for which the popover should show.
8. Open a new window.
9. Ensure that SAD/ATT popover is prompted.
10. Click the ’Not Now’ button.
11. Set a date in the future for which the SAD/ATT first banner should not show. 
12. Open a new window.
13. Ensure that no SAD/ATT banner is shown.
14. Set a date in the future for which the SAD/ATT first banner should show.
15. Ensure that SAD/ATT banner is shown.
16. Click the “x” button to dismiss.
17. Set a date in the future for which the SAD/ATT subsequent banner should not show.
18. Open a new window.
19. Ensure that SAD/ATT banner is not shown.
20. Set a date in the future for which the SAD/ATT subsequent banner should show.
21. Open a new window.
22. Ensure that SAD/ATT banner is shown.
23. Click the “Don’t ask again” button.
24. Ensure the banner is dismissed.
25. Set a date in the future for which the SAD/ATT subsequent banner should show.
26. Open a new window.
27. Ensure no banner is shown.

**Scenario 2 - Popover dismisses when Banner is presented**
1. Install the app, ensuring that the prerequisites are followed.
2. Open System Settings and click ‘Date & Time’ under ‘General'. 
3. Disable ‘Set time and date automatically’ to be able to override the current time.
4. Set a date in the future for which the SAD/ATT popover should show.
5. Open a new window. 
6. Ensure that SAD/ATT popover is prompted.
7. Do not interact with the popover.
8. Set a date in the future for which the SAD/ATT first banner should show.
9. Open a new window.
10. Ensure that the popover is dismissed and banner is shown.

**Scenario 3 - Popover dismisses when SAD/ATT state is modified from outside the popover**
1. Install the app, ensuring that the prerequisites are followed.
2. Open System Settings and click ‘Date & Time’ under ‘General'. 
3. Disable ‘Set time and date automatically’ to be able to override the current time.
4. Set a date in the future for which the SAD/ATT popover should show.
5. Open a new window. 
6. Ensure that the SAD/ATT popover is prompted.
7. Do not interact with the popover.
8. Open the App Settings or the “…” menu.
9. Add the App to Dock or Set it as default browser.
10. Ensure that the popover disappears.
11. Set a date in the future for which the “remaining” ATT/SAD banner should show.
12. Open a new window.
13. Ensure that the remaining “SAD/ATT” banner appears.

**Scenario 3 - Banner dismisses when SAD/ATT state is modified from outside the banner**
1. Install the app, ensuring that the prerequisites are followed.
2. Open System Settings and click ‘Date & Time’ under ‘General'. 
3. Disable ‘Set time and date automatically’ to be able to override the current time.
4. Set a date in the future for which the SAD/ATT popover should show.
5. Open a new window. 
6. Ensure that the SAD/ATT popover is prompted.
7. Click the “Not Now” button.
8. Set a date in the future for which the SAD/ATT first banner should show.
9. Open a new window.
10. Ensure that SAD/ATTACK banner is prompted.
11. Do not interact with the banner.
12. Open the App Settings or the “…” menu.
13. Add the App to Dock or Set it as default browser.
14. Ensure that the banner disappears.
15. Set a date in the future for which the “remaining” ATT/SAD banner should show.
16. Open a new window.
17. Ensure that the remaining “SAD/ATT” banner appears.

**Scenario 4 - Smoke test**
- Smoke test the feature to ensure SAD/ATT are set properly by confirming the prompts.

### Impact and Risks
Low: The feature is disabled.

#### What could go wrong?
We present the prompts to the user.

### Quality Considerations
<!-- 
Focus on what matters for your changes:
- What edge cases exist?
- How does this affect performance?
- What monitoring have you added?
- What documentation needs updating?
- How does this impact privacy/security?
-->

### Notes to Reviewer
<!-- Anything specific you want reviewers to focus on -->

---
###### Internal references:
[Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f) | [Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552) | [Tech Design Template](https://app.asana.com/0/59792373528535/184709971311943)

alessandroboron commented on macOS/DuckDuckGo/DefaultBrowserAndAddToDockPrompts/DefaultBrowserAndDockPromptCoordinator.swift:
```diff
@@ -106,48 +110,104 @@ final class DefaultBrowserAndDockPromptCoordinator: DefaultBrowserAndDockPrompt
+
+        // For the popover we mark it as shown when it appears on screen as we don't want to show in every windows.
+        if prompt == .popover {
+            setPopoverSeen()
```
Since the Presenter does not access the storage but it’s the coordinator’s responsibility. I’m setting this flag here before returning to value to the presenter. If we think we want to be more explicit and have the presenter ask the coordinator to do that I’m fine with that.

alessandroboron commented on macOS/DuckDuckGo/DefaultBrowserAndAddToDockPrompts/DefaultBrowserAndDockPromptCoordinator.swift:
```diff
@@ -106,48 +110,104 @@ final class DefaultBrowserAndDockPromptCoordinator: DefaultBrowserAndDockPrompt
         }
+
+        func fireActionPixel() {
+
```
Will implment Pixles in another PR

alessandroboron commented on macOS/DuckDuckGo/DefaultBrowserAndAddToDockPrompts/DefaultBrowserAndDockPromptEventMapper.swift:
```diff
@@ -0,0 +1,65 @@
+
+enum DefaultBrowserAndDockPromptEventMapper {
+
+    static let eventHandler = EventMapping<DefaultBrowserAndDockPromptEvent> { event, _, _, _ in
```
This will be used to send the debug pixels if the storage throws errors.

alessandroboron commented on macOS/DuckDuckGo/DefaultBrowserAndAddToDockPrompts/SharedInterfaces/TimerInterface.swift:
```diff
@@ -0,0 +1,51 @@
+
+import Foundation
+
+public protocol TimerInterface: AnyObject {
```
Need to find a place for this. What I mean is I put this temporary here but Ideally we should have a package that is shared between iOS and macOS. At the moment there are 3 copies of this file (1 in iOS, 1 in macOS and 1 in Onboarding package)

alessandroboron commented on macOS/DuckDuckGo/DefaultBrowserAndAddToDockPrompts/DefaultBrowserAndDockPromptPresenting.swift:
```diff
@@ -43,28 +43,26 @@ protocol DefaultBrowserAndDockPromptPresenting {
     private let coordinator: DefaultBrowserAndDockPrompt
-    private let repository: DefaultBrowserAndDockPromptStoring
-    private let featureFlagger: FeatureFlagger
+    private let statusUpdateNotifier: DefaultBrowserAndDockPromptStatusNotifying
```
I injected the `DefaultBrowserAndDockPromptStatusNotifying` here instead of the coordinator. Initially, my idea was to inject it into the coordinator, but then part of the API was reactive and part imperative (`getPromptType() -> DefaultBrowserAndDockPromptPresentationType?` ) so I injected here.

jotaemepereira reviewed (approved):
@alessandroboron This looks great, and works great too 🎉 .

Are you planning on doing translations after merging?

alessandroboron commented:
> @alessandroboron This looks great, and works great too 🎉 .
> 
> Are you planning on doing translations after merging?

Thanks @jotaemepereira !!!

I added the translation in the last commit. I also fixed the test in another commit. More info [here](https://chat.duckduckgo.com/ddg/pl/1bq36wf3eif83fqu98ufbhywxh)
